core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -1143,17 +1143,14 @@ protected Object visitLikePredicate(LikePredicate node, Object context)
                 Slice unescapedPattern = unescapeLiteralLikePattern((Slice) pattern, Optional.ofNullable((Slice) escape));
                 Type valueType = type(node.getValue());
                 Type patternType = createVarcharType(unescapedPattern.length());
-                Optional<Type> commonSuperType = typeCoercion.getCommonSuperType(valueType, patternType);
-                checkArgument(commonSuperType.isPresent(), "Missing super type when optimizing %s", node);
+                Type superType = typeCoercion.getCommonSuperType(valueType, patternType)
+                        .orElseThrow(() -> new IllegalArgumentException("Missing super type when optimizing " + node));
                 Expression valueExpression = toExpression(value, valueType);
-                Expression patternExpression = toExpression(unescapedPattern, patternType);
-                Type superType = commonSuperType.get();
                 if (!valueType.equals(superType)) {
                     valueExpression = new Cast(valueExpression, toSqlType(superType), false, typeCoercion.isTypeOnlyCoercion(valueType, superType));
                 }
-                if (!patternType.equals(superType)) {
-                    patternExpression = new Cast(patternExpression, toSqlType(superType), false, typeCoercion.isTypeOnlyCoercion(patternType, superType));
-                }
+                verify(superType instanceof VarcharType, "common super type of %s and %s is unexpectedly not varchar: %s", valueType, patternType, superType);
+                Expression patternExpression = toExpression(unescapedPattern, superType);
```

Comment:
This got simplified not to output CAST when not needed. 
`io.trino.sql.TestExpressionInterpreter#assertOptimizedEquals` requires that `ExpressionInterpreter` is more or less idempotent, so this was changed to match LiteralEncoder new behavior.